argmax for half datatype fix

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28787 argmax for half datatype fix**

Differential Revision: [D18194420](https://our.internmc.facebook.com/intern/diff/D18194420)

Fixing #28623

pbelevich committed:
argmax for half datatype fix

[ghstack-poisoned]

SsnL commented on test/test_torch.py:
```diff
@@ -8272,6 +8272,8 @@ def test_dim_reduction(self, device):
                  torch.int64,
                  torch.int32,
                  torch.int16]
+        if device.startswith('cuda'):  # 'cpu' and 'xla' do not support half
```
`device.type == 'cuda'` is the more idiomatic way to test this

pbelevich replied:
but device is a string here

pbelevich committed:
Update on "argmax for half datatype fix"




[ghstack-poisoned]

pbelevich committed:
Update on "argmax for half datatype fix"




[ghstack-poisoned]

pbelevich committed:
Update on "argmax for half datatype fix"




[ghstack-poisoned]

ngimel commented on test/test_torch.py:
```diff
@@ -8272,6 +8272,8 @@ def test_dim_reduction(self, device):
                  torch.int64,
                  torch.int32,
                  torch.int16]
+        if torch.device(device).type == 'cuda':  # 'cpu' and 'xla' do not support half
```
`self.device_type == 'cuda':`

pbelevich replied:
thanks!

ngimel reviewed (approved):
subject to minor change in test_torch

pbelevich committed:
Update on "argmax for half datatype fix"




[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/cuda/ReduceOpsKernel.cu:
```diff
@@ -169,21 +169,37 @@ void min_values_kernel_cuda(TensorIterator& iter) {
+
 void argmax_kernel_cuda(TensorIterator& iter) {
+  if (iter.dtype(1) == kHalf) {
+    argmax_kernel_cuda_impl<at::Half, float>(iter);
```
worth a comment to describe why are you doing it this way, since max is easily computable for half.

pbelevich replied:
done

pbelevich committed:
Update on "argmax for half datatype fix"


Differential Revision: [D18175646](https://our.internmc.facebook.com/intern/diff/D18175646)

[ghstack-poisoned]

pbelevich committed:
Update on "argmax for half datatype fix"


Differential Revision: [D18175646](https://our.internmc.facebook.com/intern/diff/D18175646)

[ghstack-poisoned]

pbelevich committed:
Update on "argmax for half datatype fix"


Differential Revision: [D18194420](https://our.internmc.facebook.com/intern/diff/D18194420)

[ghstack-poisoned]
